src/model/model.nit

```diff
@@ -1201,6 +1259,304 @@ abstract class MType
 	end
 end
 
+# A type that represents an operation over a set of types.
+abstract class MTypeSet[E: MType]
+	super MType
+
+	# The context in which this type was created.
+	#
+	# Used to populate `model` and look for `Object`.
+	#
+	# TODO: Remove this attribute once `as_notnull` don’t need it.
+	var mmodule: MModule
+
+	# The operands of this operation.
+	#
+	# Must not be empty. Must not be edited.
+	var operands: Set[E]
+
+	init do
+		assert operands.not_empty
+		by_operands_cache(mmodule)[operands] = self
+	end
+
+	redef fun model do return mmodule.model
+
+	redef fun location do return operands.first.location
+	# TODO: Return a more accurate location.
+
+	# Apply the same operator over the specified types.
+	#
+	# May simplify the resulting expression and not return an instance of
+	# `SELF`.
+	fun apply_to(operands: Set[MType], mmodule: MModule,
+			anchor: nullable MClassType): MType is abstract
+
+	# Cache of types according to operand sets.
+	#
+	# `mmodule` is the module in which the cache resides.
+	#
+	# SEE: `cache`
+	protected fun by_operands_cache(mmodule: MModule): Map[Set[MType], MType] is abstract
+
+	# Return the cached version of this operation over the specified set of operands.
+	#
+	# `mmodule` is the module in which the cache resides.
+	#
+	# Intended to be used internally to cache results of methods like
+	# `as_notnull`, `lookup_fixed`, `resolve_for` and `undecorate`.
+	#
+	# SEE: `apply_to`
+	protected fun cache(operands: Set[MType], mmodule: MModule): MType
+	do
+		var result = by_operands_cache(mmodule).get_or_null(operands)
+		if result == null then
```

Comment:
The cache makes more sense now, but I'm still not convinced that it improves the performance...

To put my mind at ease. Can you run a quick benchmark to check if this optimization offers a significant speedup? Comparing 2 nitc, with the cache and without the cache (by commenting out the call to `by_operands_cache`), and use them to compile the compiler or something with more type intersections. You can use `src/valgrind.sh` and compare the Ir count.